### ISSUE TITLE
C-v and M-v should do page down/up

### ISSUE THREAD

**Issue by asmeurer** [ipython/ipython#9704]
Page up/down is currently the way to cycle history without moving across multiline history items. However, in some terminals (like OS X Terminal.app), those keys don't work without binding them. For me personally, I use an MacBook Pro and I rebound Fn to Control, so I can't type those keys in the terminal (I use Karabiner to map C-v and M-v to those keys outside of the terminal, but obviously you want rebind them universally in the terminal itself). 

In emacs mode, C-v and M-v should do page down and up, respectively.

**Comment by Carreau** [comment]
2 question I like to answer first:

I guess  that can't be the case by default as Ctrl-V is past right ?
M- key is alt by default right ? Then M-AlmostAnything is used for composing characters, we likely can't take it.

**Comment by asmeurer** [comment]
Most Linux terminals that I've used bind paste to something else, since C-v is used so often. 

Regarding alt, if you're referring to the way it adds accents to characters on OS X, most people either rebind it to meta in the terminal, or use ESC. 

I'm not sure if it was clear, but these are standard emacs keybindings. It's very reasonable to expect emacs mode to use them, especially if page up/down perform such an important action. Does the vi mode use whatever vi's page up/down bindings are?

**Comment by Carreau** [comment]
> Regarding alt, if you're referring to the way it adds accents to characters on OS X, most people either rebind it to meta in the terminal, or use ESC.

Most english almost-ascii-only user. And we likely don't have the same sample of terminal users.
We did adventure to some fancy keyboard in the past (agreed on notebook, so browser...), and we ended up preventing people from typing a minus (which is useful). And having spent quite some time with a mac french keyboard, I know for example that Alt is needed to type a pipe `|` (yes I know crazy) and this is typically the kind of things that make using a default emacs unusable on such a configuration.

> I'm not sure if it was clear, but these are standard emacs keybindings.

We'll have to consider that it's also the default, and that's it's easier to bind an extra shortcut, that to unbind one.

I'm not saying we should not, and despite the Emacs-like mode, it also have to stay close to a readline-like usage.

> Most Linux terminals that I've used bind paste to something else, since C-v is used so often.

That's true, we would have to special case I guess.

Though in previous IPython (due to readline?) C-v is "insert literally next character", which might be useful as well. So we should likely be careful. 

With all these consideration I'm leaning toward :  
 - User who care can rebind [yes we need more docs and a simpler way].
 - User who will be annoyed have no clue how to deals with broken keybindings. 

A (maybe better?) solution would be to have more keymaps, a "default" one, which is limited and play it "safe", then the `emacs` and `vi` ones that are opt-in and more opinionated.

**Comment by takluyver** [comment]
Ctrl-V is already bound to paste on Windows systems, and I can imagine we may want to bind it to some form of paste on other systems at some point. I'm sure I accidentally press Ctrl-V instead of Ctrl-Shift-V in my terminal sometimes.

I'm not sure that we can capture meta key combos - PTK doesn't appear to expose any combos with a meta key modifier.

While the mode is called emacs, I think the vast majority of users are coming to it from the default readline key bindings, and are more likely to be confused than helped by closely emulating emacs. I agree that if emacs users want lots of their shortcuts, we should possibly offer a separate mode for that and rename the default. (Ping @jonathanslenders )

**Comment by asmeurer** [comment]
I didn't expect this to be so controversial. On OS X, C-v and M-v literally just enter `^V` and `^[v`, or just beep the shell. But I guess OS X is implicitly more reasonable than other systems for terminal use because the OS level shortcuts don't use control. Alt-v enters `√` with the default US English keyboard, by the way. 

If you don't want to rebind C-v and M-v (though I still think you should, for emacs mode), maybe bind C-p and C-n to what page up/down currently do? That was the first thing I tried. 

Alternate thought: it seems that part of the issue here is that emacs mode is the default, but really you want the default to be "reasonable" mode, and have a separate "emacs" mode for emacs users, just like there's a "vi" mode for vi users.

**Comment by takluyver** [comment]
Yeah, if emacs mode wasn't the default, I'd be fairly happy for whatever shortcuts to be added to it. What I care about is mainly the default experience with no config.

**Comment by Carreau** [comment]
> I didn't expect this to be so controversial.

I was, after having bitten by keyboard shortcut conflicts a few time, I'm worried now. 
Sorry about that.

**Comment by Scowley4** [comment]
Besides pg-up/pg-down, is there another way to move through "[In]"s without doing the multiline thing? I was thinking that g-j / g-k as a possible solution. 

I guess it's kinda the opposite of how vim uses g-j / g-k... but it's something.

(Is there a way I can bind this myself? Sorry if it's a dumb question, new to vi-mode in ipython)

**Comment by Carreau** [comment]
You will have to dig in the code to know which functions are available, but here is some documentation on getting started: 

https://ipython.readthedocs.io/en/stable/config/details.html#keyboard-shortcuts
